### PR TITLE
Speedup in submodel_names

### PR DESCRIPTION
**Pull request by bernie-simon**
Since submodel_names was a hotspot in the code, it seemed worth the time to optimize it. The original code updated the names in random (hash) order and then sorted. The new code updates them in their original order, doing away for the need for a sort. Also, cases where the name is None are handled as a specially. There is no value to building names like None_1, None_2, and so on. After rewriting I noticed the caching of the submodel names was incorrect, so I also changed that.

### CODE REVIEW

**Comment by pllim** [comment]
"Caching is incorrect" means this is also a bug fix?

**Comment by bernie-simon** [comment]
Since it's late afternoon here, adding comments to the code seems an agreeable way to exercise my limited mental faculties. So I will do another update.

**Comment by bernie-simon** [comment]
"Caching is incorrect" means this is also a bug fix?

The code never set the cache, so it recalculated every time the method was called. The result was correct, and thus passed the tests, but a lot of unneeded time was spent doing the calculation.

File: astropy/modeling/core.py [hunk 1]
```diff
@@ -1858,32 +1858,25 @@ def __reduce__(cls):
 
     @property
     def submodel_names(cls):
-        if cls._submodel_names is not None:
-            return cls._submodel_names
```
Comment (nden) [inline astropy/modeling/core.py]: This check should be kept. There's no need to recompute it every time.

File: astropy/modeling/core.py [hunk 2]
```diff
@@ -1858,32 +1858,46 @@ def __reduce__(cls):
...
+        if cls._submodel_names is None:
+            
+            # Duplicate names must be disambiguated.
+            seen = {}
+            names = []
+            for idx, submodel in enumerate(cls._get_submodels()):
+                name = submodel.name
+                if name is None:
+                    # No need to disambiguate None
+                    names.append(name)
```
Comment (nden) [inline astropy/modeling/core.py]: Thinking a bit more about changing how names are created for models without user set names, i.e. in master the name is set to a string of `None_#` where `#` indicates the submodel number in the adopted ordering. This PR changes it to `None`. I prefer the naming in master for two reasons. Names are consistently strings. The ordering is clear and although there's no user friendly name, the `#` shows how to get a model by indexing the compound model. Not perfect but a bit more useful than just `None`.

File: astropy/modeling/core.py (hunk 1 above)
Comment (nden) [inline astropy/modeling/core.py]: The low performance of this function comes mainly from a bug in a different place (a typo really) [here](https://github.com/astropy/astropy/pull/5741/files#diff-a619f3fef9ad65068ecd1790cebc242aL1885) fixed in #5741. Nevertheless this is a possible rewrite if it speeds up additionally.

**Comment by pllim** [comment]
@bernie-simon , please rebase to resolve conflicts and address the review comments.

**Comment by mirca** [comment]
tests are needed

**Comment by bernie-simon** [comment]
New pull request incorporates recent changes to astropy/modeling/core.py

**Comment by bsipocz** [comment]
@bernie-simon - This still (or again) has conflicts. Could you rebase again to put this on top of current upstream master?

**Comment by bernie-simon** [comment]
Will see if I can wrestle git to my will via a rebase.

**Comment by bsipocz** [comment]
@bernie-simon - I can guide you through a rebase on gitter, some time next week, if you think it would be helpful.

**Pull request closed without merging**